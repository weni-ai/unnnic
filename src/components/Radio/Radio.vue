<template>
  <div
    @click="click"
    :class="[
      'unnnic-radio-container',
      `unnnic-radio-container--${size}`,
      disabled ? 'disabled' : null,
    ]"
  >
    <unnnic-icon-svg
      class="unnnic-radio"
      :icon="icon"
      :scheme="color"
      :size="size"
    />

    <span class="label">
      <slot/>
    </span>
  </div>
</template>

<script>
import unnnicIconSvg from '../Icon.vue';

export default {
  model: {
    prop: 'globalValue',
    event: 'change',
  },

  props: {
    value: {
      type: [String, Number],
    },

    globalValue: {
      type: [String, Number],
    },

    disabled: {
      type: Boolean,
      default: false,
    },

    size: {
      type: String,
      default: 'md',
      validator(value) {
        return ['md', 'sm'].includes(value);
      },
    },
  },

  components: {
    unnnicIconSvg,
  },

  computed: {
    valueName() {
      return this.globalValue === this.value
        ? 'selected'
        : 'default';
    },

    icon() {
      return {
        selected: 'radio-selected',
        default: this.disabled ? 'radio-disable' : 'radio-default',
      }[this.valueName];
    },

    color() {
      if (this.disabled) {
        return 'brand-sec';
      }

      return this.valueName === 'selected'
        ? 'brand-weni'
        : 'neutral-cleanest';
    },
  },

  methods: {
    click() {
      if (!this.disabled) {
        this.$emit('change', this.value);
      }
    },
  },
};
</script>

<style lang="scss" scoped>
@import "../../assets/scss/unnnic.scss";

.unnnic-radio-container {
  display: inline-flex;
  align-items: center;

  &:not(.disabled) {
    cursor: pointer;
  }

  &--sm .unnnic-radio {
    margin: $unnnic-spacing-stack-nano $unnnic-spacing-inline-nano;
  }

  &--md .unnnic-radio {
    margin: $unnnic-spacing-stack-nano 0;
    margin-left: $unnnic-spacing-inline-nano;
    margin-right: $unnnic-spacing-inline-xs;
  }

  .label {
    font-family: $unnnic-font-family-secondary;
    font-weight: $unnnic-font-weight-regular;
    color: $unnnic-color-neutral-darkest;
    margin: $unnnic-spacing-stack-nano 0;
    margin-right: $unnnic-spacing-inline-nano;
  }

  &--sm .label {
    font-size: $unnnic-font-size-body-md;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
  }

  &--md .label {
    font-size: $unnnic-font-size-body-gt;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
