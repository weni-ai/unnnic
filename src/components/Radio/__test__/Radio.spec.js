import { mount } from '@vue/test-utils';
import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import Radio from '@/components/Radio/Radio.vue';
import UnnnicIcon from '@/components/Icon.vue';

describe('Radio.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(Radio, {
      props: {
        modelValue: '',
        value: 'option1',
        disabled: false,
        size: 'md',
      },
      global: {
        components: {
          UnnnicIcon,
        },
      },
      slots: {
        default: 'Option 1',
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('renders input with correct attributes', () => {
    const input = wrapper.find('input');
    expect(input.exists()).toBe(true);
    expect(input.attributes('type')).toBe('radio');
    expect(input.attributes('disabled')).toBe(undefined);
    expect(input.attributes('checked')).toBe(undefined);
  });

  test('applies disabled class when disabled prop is true', async () => {
    await wrapper.setProps({ disabled: true });
    expect(wrapper.find('.unnnic-radio__label').classes()).toContain('unnnic-radio__label--disabled');
  });

  test('icon changes based on valueName', async () => {
    await wrapper.setProps({ modelValue: 'option1' });
    const input = wrapper.find('input');
    expect(input.attributes()).toHaveProperty('checked');
  });

  test('icon and color change when disabled is true', async () => {
    await wrapper.setProps({ disabled: true });
    const input = wrapper.find('input');
    expect(input.attributes()).toHaveProperty('disabled');
  });

  test('emits update:modelValue with correct value when clicked and not disabled', async () => {
    await wrapper.find('input').trigger('change');
    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    expect(wrapper.emitted('update:modelValue')[0]).toEqual(['option1']);
  });

  test('does not emit update:modelValue when clicked and disabled', async () => {
    await wrapper.setProps({ disabled: true });
    await wrapper.trigger('click');
    expect(wrapper.emitted('update:modelValue')).toBeFalsy();
  });

  test('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
