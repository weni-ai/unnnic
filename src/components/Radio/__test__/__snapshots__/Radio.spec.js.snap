// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Radio.vue > matches the snapshot 1`] = `
"<section data-v-38c6bf5e="" class="unnnic-radio-container" size="md"><label data-v-38c6bf5e=""><input data-v-38c6bf5e="" class="unnnic-radio" type="radio" name="">
    <p data-v-38c6bf5e="" class="unnnic-radio__label"> Option 1</p>
  </label></section>"
`;
